### PR TITLE
Add ChatCraftMessage classes, support different message types

### PR DESCRIPTION
**Pull request by humphd**
In preparation for a bunch of feature work (backend db, supporting more models, file uploads, etc) I needed to refactor how the messages are stored, modelled, and displayed.

This change introduces different message types in the UI.  For example, we can now see which model made which response (green is ChatGPT, purple is GPT-4):

Screenshot 2023-05-17 at 10 46 07 AM

Internally, I've extended the langchain message types to add some metadata to each message (e.g., a unique `id` and which `model` was used, if any).  I've created our own classes (e.g., `ChatCraftMessage`, `ChatCraftAiMessage`, `ChatCraftHumanMessage`, `ChatCraftSystemMessage`) which extend all the langchain types.

I've also fixed a few things I hit along the way:

- disable the "Run" button when a message is already streaming in
- switch from `eval()` to `new Function()` to get around a rollup warning
- better display for error messages when "Run" fails

I've left the code to support the old serialization format, but upgrade messages to the new format.  Eventually we can remove this, but for now it will mean that people won't lose data as we upgrade the format on first use.

### CODE REVIEW

**Comment by humphd** [comment]
https://15d7b36f.console-overthinker-dev.pages.dev/helloworld works, but the root `/` is now broken.  I must need to specify that I want to serve the static content there explicitly...

File: src/ChatCraftMessage.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
+import { BaseChatMessage, type MessageType } from "langchain/schema";
+import { nanoid } from "nanoid";
+
+// We store ChatCraft messages as flat JSON
+export type SerializedChatCraftMessage = {
+  id: string;
+  type: MessageType;
+  model: GptModel | null;
```
Comment (tarasglek) [inline src/ChatCraftMessage.ts:8]: i would serialize model params here..maybe in a deduplicated form

**Comment by humphd** [comment]
Can you elaborate on these comments?  What you're saying is connected to what I was discussing in https://github.com/tarasglek/chatcraft.org/issues/13#issuecomment-1528852244.

Currently we don't expose any knobs for altering the [model params (temperature, etc.)](https://github.com/hwchase17/langchainjs/blob/3a6362f/langchain/src/chat_models/openai.ts#L103-L119), is that something you want to do?

What do you want in this JSON?  It would be good to pin it down now, since it will affect storage (sqlite, R2) so we don't have to support too many formats.

**Comment by humphd** [comment]
I'm going to land this so that I can build the sharing stuff on top.  I'll refine the chat/message format stuff as part of that.

**Pull request merged**